doc/lexer-modes/src/grammar.lalrpop

```diff
@@ -0,0 +1,116 @@
+// -*- mode: Rust; -*-
+//
+// This grammar parses a list of `length:value` pairs.
+//
+// In order to parse the `value` part, we have to know the length.
+// The value part can't be parsed by the grammar, but it can be lexed
+// by the lexer.  The complication is that only the grammar can detect
+// that the lexer need to change modes.  So we need some way for the
+// parser to communicate that to the lexer.  This is possible, albeit
+// a bit awkward, using some shared state.
+
+//  When we observe the length, the parser changes the lexer to
+// "literal parsing" mode.  When next called, the lexer returns a
+// token containing the next `n` bytes, and then returns to normal
+// lexing mode.
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use crate::Value;
+use crate::lexer;
+use crate::lexer::LexerMode;
+use crate::lexer::LexicalError;
+
+// Because `mode` has to be `Copy`, we have to pass a reference to the
+// `Rc<RefCell<LexerMode>>`.
+grammar<'input, 'mode>(mode: &'mode Rc<RefCell<LexerMode>>);
+
+// Zero or more whitespace separated values.
+pub List: Vec<Value> = {
+    Whitespace* <ValueWhitespace*>
+};
+
+Whitespace = {
+    SPACE, HTAB, VTAB, CR, LF, FORMFEED,
+};
+
+ValueWhitespace = {
+    <Value> Whitespace*
+}
+
+// A value is of the form: `n:bytes` where `n` is a count and bytes is
+// exactly `n` literal bytes.
+Value: Value = {
+    LiteralCount COLON <literal:LITERAL> => {
+        Value {
+            value: literal.as_bytes().to_vec()
+        }
+    }
+}
+
+// We factor this production out of the Literal production so that we
+// can change the lexer mode after we lex the COLON, but before we lex
+// the literal data.
+LiteralCount: () = {
+    <count:Decimal> => {
+        // Change the lexer to literal parsing.
+        mode.borrow_mut().literal = Some(count);
+    }
+}
+
+// A Number.
+Decimal: usize = {
```

Comment:
I don't know that it really matters too much for the purpose of documentation, but is there a reason to lex the counts by digit and then reconstruct numbers at the parser level rather than when the lexer isn't in literal mode just taking all the digits up to the colon as a single token?

Reply:
This example is a distilled version of a [sexp parser](https://datatracker.ietf.org/doc/html/draft-rivest-sexp-09), [which I wrote](https://gitlab.com/sequoia-pgp/sequoia/-/commit/464d289daf0888cd4eeaa02a23ec1cbbecbd5686).  As far as I can tell, when parsing sexp expressions, the lexer doesn't have enough information to decide how to parse the digits.  In this example, it does, and if you want to change it, that's fine by me.

Reply:
Okay, I'm fine with leaving it as is.  We can always update it post-merge if desired.